components/notebook-controller/pkg/culler/culler.go

```diff
@@ -156,51 +173,145 @@ func getNotebookApiStatus(nm, ns string) *NotebookStatus {
 		return nil
 	}
 
-	status := new(NotebookStatus)
-	err = json.NewDecoder(resp.Body).Decode(status)
+	var kernels []KernelStatus
+
+	err = json.NewDecoder(resp.Body).Decode(&kernels)
 	if err != nil {
-		log.Info(fmt.Sprintf(
-			"Error parsing the JSON response for Notebook %s/%s", nm, ns),
-			"error", err)
+		log.Error(err, "Error parsing JSON response for Notebook API Kernels.")
 		return nil
 	}
 
-	return status
+	return kernels
 }
 
-func notebookIsIdle(nm, ns string, status *NotebookStatus) bool {
-	// Being idle means that the Notebook can be culled
-	if status == nil {
+func allKernelsAreIdle(kernels []KernelStatus, log logr.Logger) bool {
+	// Iterate on the list of kernels' status.
+	// If all kernels are on execution_state=idle then this function returns true.
+	log.Info("Examining if all kernels are idle")
+
+	if kernels == nil {
 		return false
 	}
 
-	lastActivity, err := time.Parse(time.RFC3339, status.LastActivity)
-	if err != nil {
-		log.Info(fmt.Sprintf("Error parsing time for Notebook %s/%s", nm, ns),
-			"error", err)
-		return false
+	for i := 0; i < len(kernels); i++ {
+		if kernels[i].ExecutionState != KERNEL_EXECUTION_STATE_IDLE {
+			log.Info("Not all kernels are idle")
+			return false
+		}
 	}
+	log.Info("All kernels are idle")
+	return true
+}
+
+// Update LAST_ACTIVITY_ANNOTATION
+func UpdateNotebookLastActivityAnnotation(meta *metav1.ObjectMeta) {
+	log := log.WithValues("notebook", getNamespacedNameFromMeta(*meta))
+	if meta == nil {
+		log.Info("Metadata is Nil. Can't update Last Activity Annotation.")
+		return
+	}
+
+	log.Info("Updating the last-activity annotation.")
+	nm, ns := meta.GetName(), meta.GetNamespace()
 
-	timeCap := lastActivity.Add(getMaxIdleTime())
-	if time.Now().After(timeCap) {
+	// No last-activity found in the CR. Setting to Now()
+	if _, ok := meta.GetAnnotations()[LAST_ACTIVITY_ANNOTATION]; !ok {
+		t := createTimestamp()
+		log.Info(fmt.Sprintf("No last-activity found in the CR. Setting to %s", t))
+
+		if len(meta.GetAnnotations()) == 0 {
+			meta.SetAnnotations(map[string]string{})
+		}
+		meta.Annotations[LAST_ACTIVITY_ANNOTATION] = t
 		return true
```

Comment:
@athamark can you change this to a single `return`? The function's signature does not return a value

Reply:
@kimwnasptd thanks for the feedback. I corrected this, I pushed a commit that resolves this, you should be able to see it as `review: UpdateNotebookLastActivityAnnotation()`.